prep 2.0.0 (please review)

Phil-Friderici opened the pull request:
Uses data types, adds puppet strings inline docs, uses facts hash instead of lsb-facts.

Other changes:
- Removed support for stringified booleans and integers
- Removed option 'held' from $updatesd_package_ensure as it makes no sense on yum.
- All yum.conf specific parameters will only be added when their value is not an empty string or array.
- Changed default value for keepcache in main section of yum.conf to '1' (true) to follow yum standard.
- Changed default value for gpgcheck in main section of yum.conf to '0' (false) to follow yum standard.
- Changed default value for tolerant in main section of yum.conf to '0' (false) to follow yum standard.
- Changed default value for plugins in main section of yum.conf to '0' (false) to follow yum standard.
- Changed default value for metadata_expire in main section of yum.conf to '6h' to follow yum standard.
- distroverpkg now also allows free string texts.
- exclude does now only suppport arrays. Support for strings was dropped.
- Changed default value for yum::repo::gpgcheck in main section of yum.conf to '0' (false) to follow yum standard.
- Removed non existing yum::repo::priority parameter.
- Removed dependency for ghoneycutt/common module

ghoneycutt commented on manifests/init.pp:
```diff
@@ -2,111 +2,112 @@
+  Boolean $manage_repos                       = false,
+  String $repos_d_owner                       = 'root',
+  String $repos_d_group                       = 'root',
+  Pattern[/^[0-7]{4}$/] $repos_d_mode         = '0755',
```
Here's a better pattern

https://github.com/puppetlabs/puppetlabs-stdlib/commit/99bbe0ba5ea5c6141b3b4435bd2b0e7a77dd6c43#diff-b955865310c37e5eaa511c3abe074bdcR1

Hopefully that get's put in a tagged release soon and we can just use Stdlib::Filemode

Phil-Friderici replied:
Have seen that before, should we wait until stdlib releases this ?

ghoneycutt replied:
Nah, let's not wait. We can use the pattern from there though.

ghoneycutt replied:
4.22.0 has `Stdlib::Filemode` now. Let's use that.

Phil-Friderici replied:
changed

ghoneycutt commented on manifests/init.pp:
```diff
@@ -2,111 +2,112 @@
       default: { $repos_real = $repos }
     }
     validate_hash($repos_real)
     create_resources('yum::repo',$repos_real)
```
we should stop using create_resources and do this the Puppet4 way with lambda's.

https://github.com/sensu/sensu-puppet/blob/master/manifests/init.pp#L521-L525

Phil-Friderici replied:
Can work on this in the next days.

Phil-Friderici replied:
changed

ghoneycutt commented on manifests/repo.pp:
```diff
@@ -5,126 +5,160 @@
+  Optional[String] $username                = undef,
+  Optional[String] $password                = undef,
+  Optional[String] $description             = $name,
+  String $environment                       = "${::environment}", # lint:ignore:only_variable_string # stringification needed for spec testing on Puppet 3.8 & 4.3.0
```
Do we still need the stringification noted in the comment?

Phil-Friderici replied:
nope, removed

ghoneycutt commented on manifests/repo.pp:
```diff
@@ -5,126 +5,160 @@
-#  yum::repo { 'redhat-base':
-#    gpgcheck => '1',
-#  }
+# === Parameters
```
This should be under `@summary` Check this out

https://github.com/sensu/sensu-puppet/blob/master/manifests/init.pp#L1-L4

see it in action at

http://sensu.github.io/sensu-puppet/doc/puppet_classes/sensu.html

Phil-Friderici replied:
Thanks, didn't know that. Updated accordingly.

ghoneycutt commented on spec/classes/init_spec.rb:
```diff
@@ -3,10 +3,6 @@
   mandatory_facts = {
     :fqdn                => 'no-hiera-data.example.local',
     :test                => 'no-hiera-data',
-    :environment         => 'rp_env',
```
we dont need this magic environment anymore? great!

Phil-Friderici replied:
it moved to ./spec/spec_helper.rb as it is needed by `yum::repo`. But it haven't to be 'rp_env' anymore :)

ghoneycutt commented on .travis.yml:
```diff
@@ -20,24 +20,24 @@ script: 'SPEC_OPTS="--format documentation" bundle exec rake validate lint spec'
+    - rvm: 2.1.9
+      env: PUPPET_GEM_VERSION="~> 4.8.0"
+    - rvm: 2.1.9
+      env: PUPPET_GEM_VERSION="~> 4.9.0"
```
Let's get rid of all the minor versions of 4 and just test `~> 4` since if you're running v4, you should be at the latest release.

Phil-Friderici replied:
done

ghoneycutt commented:
Seriously impressive work! Thank you!

Phil-Friderici commented on spec/classes/init_spec.rb:
```diff
@@ -226,60 +226,60 @@
-        :invalid => ['../invalid', %w(array), { 'ha' => 'sh' }, 3, 2.42, true, nil],
-        :message => 'expects.*Variant\[Stdlib::Windowspath.*Stdlib::Unixpath',
+        :invalid => ['../invalid', %w(array), { 'ha' => 'sh' }, 3, 2.42, false, nil],
+        :message => 'expects a (match for|match for Stdlib::Absolutepath =|Stdlib::Absolutepath =) Variant\[Stdlib::Windowspath.*Stdlib::Unixpath', # Puppet (4.x|5.0 & 5.1|5.x)
```
They did it again, changing error messages for minor versions.
Puppet 5.0 & 5.1 error messages are different from >= 5.2.... tzzzz....

ghoneycutt commented on manifests/init.pp:
```diff
@@ -1,112 +1,693 @@
+  Variant[Enum[''],Stdlib::Absolutepath] $persistdir      = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $pluginconfpath  = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $pluginpath      = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $sslcacert       = '',
```
Why the empty quotes instead of `undef` ?

Phil-Friderici replied:
You are not able to set undef/nil from hiera. With an empty string you can override these settings back to their defaults even if you have set it on another hiera level.

Phil-Friderici replied:
Actually I try to avoid the strange case of using an undef boolean to trigger a functionality.
A boolean should only be a boolean and nothing else.

ghoneycutt replied:
+1 a boolean should only be a boolean

You can specify `undef` in yaml by using `~`.

```yaml
foo::bar: ~
```

ghoneycutt replied:
For path's I think we want the following

```puppet
Optional[Stdlib::Absolutepath] $path  = undef,
Optional[Stdlib::Absolutepath] $path2 = '/path/to/file',
```



https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#optional

Phil-Friderici replied:
The intention was to always use the empty strings as trigger to avoid adding a param to yum.conf.
Otherwise we end up in some times undef and sometimes empty strings as triggers.

Think using one pattern only is more predictible.

Phil-Friderici commented:
@ghoneycutt 
**/!\ need feedback /!\\**
Unsure if the concatenation of baseurl & gpgkey is still needed.
The default result with three backslashes looks somewhat suspicious. (see repo_spec.rb line 99-109)

We have username and password as specific parameters for basic auth in the repository file already.

I am sure this part needs more refactoring, just not sure in which direction.

Phil-Friderici commented:
Now parameters for repositoriy files are feature complete too.

From my perspective we just need to rethink if we want to continue the concatenation for baseurl & gpgkey.

ghoneycutt commented on manifests/server.pp:
```diff
@@ -1,28 +1,42 @@
+  String $gpg_keys_path             = 'keys', # gpg_keys_path is relative to $docroot, ${docroot}/${gpg_keys_path}
+  String $gpg_user_name             = 'Root',
+  String $yum_server                = 'yum',
+  String $yum_server_http_listen_ip = $::ipaddress,
```
Check out https://github.com/thrnio/puppet-ip for validating the IP address

Phil-Friderici replied:
Really want to add another dependency just to check IP addesses ?

ghoneycutt replied:
yeah, we're already using it elsewhere and it seems to be the right way for network validation.

ghoneycutt commented:
@Phil-Friderici the yum::repo defined type was initially created to be used with yum::server so that you could easily host gpg keys and repo's you managed yourself.

Now that yum::repo is feature complete, perhaps we no longer need this functionality. What do you think about getting rid of all special logic and params that essentially build paths to the known naming scheme for the gpg keys and baseurl and instead just document how you can use the defined type to specify the most common usage with baseurl and gpg enabled.

Phil-Friderici commented:
Ok, removed the concatination of $gpgkey & $baseurl. Will try to document the use cases for them.

Phil-Friderici commented:
Both functionalities of ghoneycutt/common were only used in one case each. To make it easy I have copied the functionality directly into the classes using it.

Phil-Friderici commented:
still need to work on "use https://github.com/thrnio/puppet-ip for validate IP addresses"

Phil-Friderici commented:
Strange, added puppet-ip to the dependencies but the data types are not visible:
```
      Puppet::PreformattedError:
        Evaluation Error: Resource type not found: IP::Address at /root/puppet/puppet-module-yum/spec/fixtures/modules/yum/manifests/server.pp:33:3
```
Do you know any existing use case for puppet-ip that I can peek into ?

ghoneycutt commented on metadata.json:
```diff
@@ -12,6 +12,7 @@
     {"name":"puppetlabs/firewall","version_requirement":">= 1.8.1 < 2.0.0"},
-    {"name":"puppetlabs/concat","version_requirement":">= 2.2.0 < 3.0.0"}
+    {"name":"puppetlabs/concat","version_requirement":">= 2.2.0 < 3.0.0"},
+    {"name":"thrnio/puppet-ip","version_requirement":">= 1.0.0 < 2.0.0"}
```
should be `thrnio/ip`

that's what's causing travis to fail

Phil-Friderici replied:
Thanks @ghoneycutt, now it works like a charm :)

Phil-Friderici commented:
We got hit by a bug in thrnio/ip when running on Puppet 5.2. PR https://github.com/ghoneycutt/puppet-module-yum/issues/23 seems to fix it but didn't got merged since September.

ghoneycutt commented on manifests/repo.pp:
```diff
@@ -1,148 +1,239 @@
+# @param yum_repos_d_path
+#   Specify the path of the directory for yum repository files.
+#
+# @param gpgkey_local_path
```
this does not seem to be used, also I can't find documentation for it

ghoneycutt replied:
ah this seems to be part of the old module where it would build a URL for you and download and install the gpg key for you. We want to remove all of that old functionality and just support whatever the man page says is supported. We can then document what steps are needed to download and install a key with `yum::rpm_gpg_key`

Phil-Friderici replied:
With "Deprecate GPG key download in repo define" the old functionality is removed completely.
Have added `$rpm_gpg_keys` and `$rpm_gpg_keys_hiera_merge` for easy GPG key management.
Documentation is updated accordingly.

ghoneycutt commented on manifests/repo.pp:
```diff
@@ -1,148 +1,239 @@
+#   When empty, it will not be present.
+#
+# @param gpgkey
+#   gpgkey setting in the repository file.  Accepts HTTP/HTTPS URLs.
```
This should also accept `file://` URL's.

This should accept multiple lines. The format is like this


```
gpgkey=file:///key1
       file:///key2
```

Notice there is no equal on the other lines. Perhaps use

```ruby
.join("\n       ")
```

Phil-Friderici replied:
fixed this also for baseurl which do also allow the same URLs in "Allow ftp and file URLs for baseurl and gpgkey".

For easier spec testing I have choosen to have two spaces in both cases:
```
.join("\n  ")
```

ghoneycutt replied:
Thanks! I was really surprised that the config file could be line space delimited instead of comma delimited for certain options. Very odd.

ghoneycutt commented:
Just noticed that the puppetlabs-release-pc1 package sets up the repo to use multiple gpg keys.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -686,6 +696,20 @@
 
+  if $rpm_gpg_keys != undef {
+    case $rpm_gpg_keys_hiera_merge {
+      true:    { $rpm_gpg_keys_real = hiera_hash('yum::rpm_gpg_keys') }
```
hiera functions are deprecated in favor of `lookup()`. This function can set if the lookup should be a merge like hiera_hash() is as well and validate the data so we don't have to use the validate_hash() function.

https://puppet.com/docs/puppet/5.3/hiera_automatic.html

Phil-Friderici replied:
Changed in "Use lookup() in favor of hiera_hash()".

Phil-Friderici commented:
Regarding thrnio/ip: we could exclude support for Puppet 5.2 until it got fixed.

ghoneycutt commented:
I sent an email to thrnio asking for merge access, hopefully he responds soon.

It does still work for Puppet 5.3 and this module is going to support the latest release of Puppet 5, so 5.2 being broken is acceptable.

Phil-Friderici commented:
I'll update the Travis configuration to accpet failing on Puppet 5.2 temporarily.

ghoneycutt commented:
I've been just testing the latest release for 4 and 5. We supported minor releases for Puppet 3 and that was a good exercise in maintaining software that worked for all those versions, though as we rework things, I don't want to go down that path again. It was a ton of work and constant updates to maintain compatibility with software releases which are supposed to be safe to upgrade through.

Phil-Friderici commented:
You are right, testing and supporting all old versions was a lot of work. On the other side it was a good training to write code that works everywhere. And I also think about unit tests on different Rubies :)

ghoneycutt commented:
Instead of using `''` we should be using `undef`. Could you please refactor the data types for that. I believe you change `''` in the templates to `nil`.

ghoneycutt commented:
I'm going to start working on a vagrant setup so we can test this functionally.

ghoneycutt commented on manifests/updatesd.pp:
```diff
@@ -1,15 +1,30 @@
+  Enum['absent', 'latest', 'present', 'purged'] $updatesd_package_ensure = 'absent',
+  String $updatesd_service                                               = 'yum-updatesd',
+  Enum['running', 'stopped'] $updatesd_service_ensure                    = 'stopped',
+  Enum['false', 'manual', 'mark', 'true'] $updatesd_service_enable       = 'false', # lint:ignore:quoted_booleans
```
Perhaps instead of us doing type validation for enable and ensure, we should just check that a string or boolean is being passed. Then if Puppet adds another possible value for `enable` or whatever, it would be supported. If you pass the wrong thing, Puppet will already error.

Phil-Friderici replied:
For ensure on services we can use [Stdlib::Ensure::Service](https://github.com/puppetlabs/puppetlabs-stdlib#stdlibensureservice).

ghoneycutt commented on manifests/repo.pp:
```diff
@@ -1,148 +1,235 @@
-# == Define: yum::repo
+# @summary Manage individual yum repo files in /etc/yum.repos.d
```
Needs to `include ::yum`

Phil-Friderici replied:
`include ::yum` causes a duplicate resource declaration:
```
     Failure/Error: include ::yum

     Puppet::PreformattedError:
       Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Resource Statement, Duplicate declaration: Exec[clean_yum_cache] is already declared in file :2; cannot redeclare at /root/puppet/puppet-module-yum/spec/fixtures/modules/yum/manifests/init.pp:737 at /root/puppet/puppet-module-yum/spec/fixtures/modules/yum/manifests/init.pp:737:3 at line 7 on node learning.puppetlabs.vm
```

ghoneycutt commented:
Ryan merged the outstanding PR's from thrnio/ip and made me a collaborator on the project! I'm also attempting to get the functionality merged into stdlib at https://github.com/puppetlabs/puppetlabs-stdlib/pull/872

Phil-Friderici commented:
Having issues to override parameters that have default values with undef in the tests. Will continue to try to find a workaround without removing the test cases.

Btw: now using thrnio/ip v1.0.1.

ghoneycutt commented:
@Phil-Friderici made some changes and sent a PR against your branch at https://github.com/Phil-Friderici/puppet-module-yum/pull/4

Phil-Friderici commented:
- changed all default cases from '' to undef.
  removed some tests in https://github.com/ghoneycutt/puppet-module-yum/pull/24/commits/82328d2cf7daaeea000f112b6e0bfa176a77868a, didn't found a way to override the defaults with undef/nil
- using fixed version 1.0.1 of thrnio/ip
- removed enum validations for ensure and enable parameters
- merged PR for vagrant setup
